pkg/pipeline-selector/pipeline-selector.go

```diff
@@ -0,0 +1,152 @@
+	devfile "github.com/redhat-appstudio/application-service/pkg/devfile"
+	buildappstudiov1alpha1 "github.com/redhat-appstudio/build-service/api/v1alpha1"
+)
+
+// SelectPipelineForComponent evaluates given list of pipeline selectors aginst specified component
+// to find the build pipeline for the component.
+// The first match is returned.
+func SelectPipelineForComponent(component *appstudiov1alpha1.Component, selectors []buildappstudiov1alpha1.InitialBuildPipelineSelector) (*tektonapi.PipelineRef, []tektonapi.Param, error) {
+	selectionParameters, err := getPipelineSelectionParametersForComponent(component)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, selector := range selectors {
+		if buildPipelineRef, buildPipelineAdditionalParams := findMatchingPipeline(selectionParameters, &selector); buildPipelineRef != nil {
+			return buildPipelineRef, buildPipelineAdditionalParams, nil
+		}
+	}
+	return nil, nil, nil
+}
+
+// getPipelineSelectionParametersForComponent returns build parameters of the given component
+// to use when matching to the build pipelines selection logic.
+func getPipelineSelectionParametersForComponent(component *appstudiov1alpha1.Component) (*buildappstudiov1alpha1.WhenCondition, error) {
+	parameters := &buildappstudiov1alpha1.WhenCondition{}
+
+	parameters.ComponentName = component.GetName()
+	parameters.Annotations = component.GetAnnotations()
+	parameters.Labels = component.GetLabels()
+
+	devfileData, err := devfile.ParseDevfileModel(component.Status.Devfile)
+	if err != nil {
+		return nil, err
+	}
+
+	devfileMetadata := devfileData.GetMetadata()
+
+	parameters.Language = devfileMetadata.Language
+	parameters.ProjectType = devfileMetadata.ProjectType
+
+	if dockerfile, err := devfile.SearchForDockerfile([]byte(component.Status.Devfile)); err == nil && dockerfile != nil {
+		parameters.DockerfileRequired = buildappstudiov1alpha1.Bool_true
+	} else {
+		parameters.DockerfileRequired = buildappstudiov1alpha1.Bool_false
+	}
+
+	return parameters, nil
+}
+
+// findMatchingPipeline evaluates given selectors chain against component parameters.
+// The first match is returned.
+func findMatchingPipeline(selectionParameters *buildappstudiov1alpha1.WhenCondition, selectors *buildappstudiov1alpha1.InitialBuildPipelineSelector) (*tektonapi.PipelineRef, []tektonapi.Param) {
+	for _, pipelineSelector := range selectors.Spec.Selectors {
+		if conditionsMatchComponentParameters(&pipelineSelector.WhenConditions, selectionParameters) {
+			return &pipelineSelector.PipelineRef, pipelineSelector.PipelineParams
+		}
+	}
+	return nil, nil
+}
+
+// conditionsMatchComponentParameters evaluates given pipeline selector against component parameters.
+// In other words, checks if given pipeline can build the component (according to what the pipeline conditions say).
+func conditionsMatchComponentParameters(pipeline, component *buildappstudiov1alpha1.WhenCondition) bool {
```

Comment:
Would it be good to specify type for argument `pipeline`?

Reply:
Both parameters have the same type